#include "clickerwindow.h"
#include "mainwindow.h"
#include "ui_clickerwindow.h"
#include <QMessageBox>
#include <QTimer>
#include <QTime>

ClickerWindow::ClickerWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::ClickerWindow)
    , m_clickCount(0)
    , m_reward(0)
    , m_remainingTime(30)
{
    ui->setupUi(this);

    QPixmap bkgnd(":/game/images/resources/clicker_background.png");
    bkgnd = bkgnd.scaled(size(), Qt::KeepAspectRatioByExpanding, Qt::SmoothTransformation);

    QPalette palette;
    palette.setBrush(QPalette::Window, QBrush(bkgnd));
    this->setPalette(palette);

    // –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è —Ç–∞–π–º–µ—Ä–∞
    m_timer = new QTimer(this);
    connect(m_timer, &QTimer::timeout, this, &ClickerWindow::updateTimer);

    // –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –≤—Ä–µ–º–µ–Ω–∏
    m_time.start();
    m_timer->start(1000);

    ui->labelClicks->setText("–ö–ª–∏–∫–∏: 0");
    ui->labelReward->setText("–ù–∞–≥—Ä–∞–¥–∞: 0 –º–æ–Ω–µ—Ç ü™ô");
    ui->labelTimeRemaining->setText("–û—Å—Ç–∞–≤—à–µ–µ—Å—è –≤—Ä–µ–º—è: 30 —Å–µ–∫—É–Ω–¥");
}

ClickerWindow::~ClickerWindow()
{
    delete ui;
}

void ClickerWindow::on_clickButton_clicked()
{
    m_clickCount++;
    ui->labelClicks->setText("–ö–ª–∏–∫–∏: " + QString::number(m_clickCount));

    // –û–±–Ω–æ–≤–ª—è–µ–º –Ω–∞–≥—Ä–∞–¥—É
    if (m_clickCount >= 120) {
        m_reward = 60 + (m_clickCount - 50) * 2; // –ó–∞ –∫–∞–∂–¥—ã–π —Å–ª–µ–¥—É—é—â–∏–π –±–ª–æ–∫ –∏–∑ 10 –∫–ª–∏–∫–æ–≤ +10 –º–æ–Ω–µ—Ç
    } else if (m_clickCount >= 90) {
        m_reward = 25;
    } else if (m_clickCount >= 60) {
        m_reward = 5;
    }

    ui->labelReward->setText("–ù–∞–≥—Ä–∞–¥–∞: " + QString::number(m_reward) + " –º–æ–Ω–µ—Ç ü™ô");

    if (m_time.elapsed() >= 30000) {
        endGame();
    }
}

void ClickerWindow::updateTimer()
{

    // –£–º–µ–Ω—å—à–∞–µ–º –æ—Å—Ç–∞–≤—à–µ–µ—Å—è –≤—Ä–µ–º—è
    m_remainingTime = 30 - m_time.elapsed() / 1000;

    // –û–±–Ω–æ–≤–ª—è–µ–º —Ç–µ–∫—Å—Ç–æ–≤–æ–µ –ø–æ–ª–µ –¥–ª—è –æ—Å—Ç–∞–≤—à–µ–≥–æ—Å—è –≤—Ä–µ–º–µ–Ω–∏
    ui->labelTimeRemaining->setText("–û—Å—Ç–∞–≤—à–µ–µ—Å—è –≤—Ä–µ–º—è: " + QString::number(m_remainingTime) + " —Å–µ–∫—É–Ω–¥");
    if (m_time.elapsed() >= 30000) {
        endGame();
    }
}

void ClickerWindow::endGame()
{
    m_timer->stop();

    QMessageBox::information(this, "–ò–≥—Ä–∞ –æ–∫–æ–Ω—á–µ–Ω–∞", "–í—ã —Å–¥–µ–ª–∞–ª–∏ " + QString::number(m_clickCount) + " –∫–ª–∏–∫–æ–≤.\n–ù–∞–≥—Ä–∞–¥–∞: " + QString::number(m_reward) + " –º–æ–Ω–µ—Ç ü™ô");

    // –û–±–Ω–æ–≤–ª—è–µ–º –±–∞–ª–∞–Ω—Å –≤ –≥–ª–∞–≤–Ω–æ–º –æ–∫–Ω–µ
    MainWindow *mainWindow = qobject_cast<MainWindow*>(parent());
    if (mainWindow) {
        mainWindow->updateBalance(m_reward);
    }

    this->close();
}
